sqlite.go

```diff
@@ -0,0 +1,530 @@
+		str = strings.TrimSuffix(str, ".000")
+		str = strings.TrimSuffix(str, ":00")
+		return true, s.stmt.BindText64(ordinal, str)
+	default:
+		return false, nil
+	}
+}
+
+type rows struct {
+	stmt   *stmt
+	closed bool
+
+	colNames []string // filled on call to Columns
+
+	// Filled on first call to Next.
+	colTypes     []sqliteh.ColumnType
+	colDeclTypes []string
+}
+
+func (r *rows) Columns() []string {
+	if r.closed {
+		panic("Columns called after Rows was closed")
+	}
+	if r.colNames == nil {
+		r.colNames = make([]string, r.stmt.stmt.ColumnCount())
+		for i := range r.colNames {
+			r.colNames[i] = r.stmt.stmt.ColumnName(i)
+		}
+	}
+	return append([]string{}, r.colNames...)
+}
+
+func (r *rows) Close() error {
+	if r.closed {
+		return errors.New("sqlite rows result already closed")
+	}
+	r.closed = true
+	if err := r.stmt.stmt.Reset(); err != nil {
+		return r.stmt.reserr("Rows.Close(Reset)", err)
+	}
+	return r.stmt.reserr("Rows.Close(ClearBindings)", r.stmt.stmt.ClearBindings())
+}
+
+func (r *rows) Next(dest []driver.Value) error {
+	if r.closed {
+		return errors.New("sqlite rows result already closed")
+	}
+	hasRow, err := r.stmt.stmt.Step()
+	if err != nil {
+		return err
+	}
+	if !hasRow {
+		return io.EOF
+	}
+
+	if r.colTypes == nil {
+		colCount := r.stmt.stmt.ColumnCount()
+		r.colTypes = make([]sqliteh.ColumnType, colCount)
+		r.colDeclTypes = make([]string, colCount)
+
+		for i := range r.colTypes {
+			r.colTypes[i] = r.stmt.stmt.ColumnType(i)
+			r.colDeclTypes[i] = r.stmt.stmt.ColumnDeclType(i)
+		}
+	}
+
+	for i := range dest {
+		switch r.colDeclTypes[i] {
+		case "DATETIME", "DATE":
+			switch r.colTypes[i] {
+			case sqliteh.SQLITE_INTEGER:
+				v := r.stmt.stmt.ColumnInt64(i)
+				dest[i] = time.Unix(v, 0)
+			case sqliteh.SQLITE_FLOAT:
+				dest[i] = r.stmt.stmt.ColumnDouble(i)
+			case sqliteh.SQLITE_TEXT:
+				v := r.stmt.stmt.ColumnText(i)
+				t, err := time.Parse(TimeFormat, v)
+				if err != nil {
+					return fmt.Errorf("cannot parse time from column %d: %v", i, err)
+				}
+				dest[i] = t
+			}
+			continue
+		}
+		switch r.colTypes[i] {
+		case sqliteh.SQLITE_INTEGER:
+			val := r.stmt.stmt.ColumnInt64(i)
+			if strings.EqualFold(r.colDeclTypes[i], "BOOLEAN") {
+				dest[i] = val > 0
+			} else {
+				dest[i] = val
+			}
+		case sqliteh.SQLITE_FLOAT:
+			dest[i] = r.stmt.stmt.ColumnDouble(i)
+		case sqliteh.SQLITE_TEXT:
+			dest[i] = r.stmt.stmt.ColumnText(i)
+		case sqliteh.SQLITE_BLOB:
+			dest[i] = r.stmt.stmt.ColumnBlob(i)
+		case sqliteh.SQLITE_NULL:
+			dest[i] = nil
+		}
+	}
+	return nil
+}
+
+// Error is an error produced by SQLite.
+type Error struct {
+	Code  sqliteh.Code // SQLite extended error code (SQLITE_OK is an invalid value)
+	Loc   string       // method name that generated the error
+	Query string       // original SQL query text
+	Msg   string       // value of sqlite3_errmsg, set sqlite.ErrMsg = true
+}
+
+func (err Error) Error() string {
+	str := "sqlite"
```

Comment:
If you're concerned about allocating extra times during the error path (as above, see ErrMsg), which I approve of, then should probably use a strings.Builder here.